Userland/DevTools/UserspaceEmulator/MallocTracer.cpp

```diff
@@ -349,19 +354,51 @@ bool MallocTracer::is_reachable(const Mallocation& mallocation) const
             return IterationDecision::Continue;
 
         size_t pointers_in_region = region.size() / sizeof(u32);
+
         for (size_t i = 0; i < pointers_in_region; ++i) {
             auto value = region.read32(i * sizeof(u32));
-            if (value.value() == mallocation.address && !value.is_uninitialized()) {
+            auto other_address = value.value();
+            if (!value.is_uninitialized() && m_memory_graph.contains(value.value())) {
 #if REACHABLE_DEBUG
-                reportln("mallocation {:p} is reachable from region {:p}-{:p}", mallocation.address, region.base(), region.end() - 1);
+                reportln("region/mallocation {:p} is reachable from region {:p}-{:p}", other_address, region.base(), region.end() - 1);
 #endif
-                reachable = true;
-                return IterationDecision::Break;
+                m_memory_graph.find(other_address)->value.is_reachable = true;
+                reachable_mallocations.append(other_address);
             }
         }
         return IterationDecision::Continue;
     });
-    return reachable;
+
+    // Propagate reachability
+    // There are probably better ways to do that
+    Vector<FlatPtr> visited = {};
+    for (size_t i = 0; i < reachable_mallocations.size(); ++i) {
+        auto reachable = reachable_mallocations.at(i);
+        if (visited.contains_slow(reachable))
+            continue;
+        visited.append(reachable);
+        auto& mallocation_node = m_memory_graph.find(reachable)->value;
+
+        if (!mallocation_node.is_reachable) {
+            mallocation_node.is_reachable = true;
+            m_memory_graph.set(reachable, mallocation_node);
+        }
+
+        for (auto& edge : mallocation_node.edges_from_node) {
+            reachable_mallocations.append(edge);
+        }
+    }
+}
+
+void MallocTracer::dump_memory_graph(MemoryGraph graph)
```

Comment:
You're passing the graph here by value, which (in this case) results in a copy of the entire memory graph.

Few options to avoid that:

1) Pass by const reference (generally a good idea)
2) don't pass anything and use m_memory_graph directly (good idea in this situation)
3) re-invent your MemoryGraph struct and add a `dump()` method to it. Or `show()`, but I think the pattern in serenity is more towards `dump()`. (maybe excessive, there's not any other specific operations on that graph that are particularly interesting at the moment)

Reply:
It would be nice to have a generic memory dumping utility, if there arises the need for other memory graphs. For nopw though I just use the m_memory_graph.